src/provider/database.rs

```diff
@@ -6,21 +6,880 @@ use crate::{
+    }
+}
+
+impl tokio::io::AsyncRead for AsyncDataReader {
+    fn poll_read(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<tokio::io::Result<()>> {
+        Pin::new(&mut self.0).poll_read(cx, buf)
+    }
+}
+
+impl tokio::io::AsyncSeek for AsyncDataReader {
+    fn start_seek(mut self: std::pin::Pin<&mut Self>, position: io::SeekFrom) -> io::Result<()> {
+        Pin::new(&mut self.0).start_seek(position)
+    }
+
+    fn poll_complete(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<io::Result<u64>> {
+        Pin::new(&mut self.0).poll_complete(cx)
+    }
+}
+
+impl BlobData {
+    fn reader(&self, default_path: impl Fn() -> PathBuf) -> io::Result<DataReader> {
+        Ok(match self {
+            Self::Inline(data) => DataReader::Memory(std::io::Cursor::new(data.clone())),
+            Self::File(custom_path) => DataReader::File(match custom_path {
+                Some(path) => std::fs::File::open(path)?,
+                None => std::fs::File::open(default_path())?,
+            }),
+        })
+    }
+
+    fn async_reader(
+        &self,
+        default_path: impl Fn() -> PathBuf,
+    ) -> impl Future<Output = io::Result<AsyncDataReader>> + 'static {
+        match self {
+            Self::Inline(data) => {
+                futures::future::ready(Ok(AsyncDataReader::memory(data.clone()))).left_future()
+            }
+            Self::File(custom_path) => {
+                let path = match custom_path {
+                    Some(path) => path.clone(),
+                    None => default_path(),
+                };
+                async move {
+                    let file = tokio::fs::File::open(path).await?;
+                    Ok(AsyncDataReader(Either::Right(file)))
+                }
+                .right_future()
+            }
+        }
+    }
+
+    fn exists(&self, default_path: impl Fn() -> PathBuf) -> bool {
+        match self {
+            Self::Inline(_) => true,
+            Self::File(Some(path)) => path.exists(),
+            Self::File(None) => default_path().exists(),
+        }
+    }
+}
+
+/// A bao file consists of outboard and content.
+///
+/// Either of them can stored inline or in a file in the file system.
+///
+/// Under some circumstances, we can have multiple content files. We will
+/// however not keep track of multiple outboard files.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct BaoFile {
+    outboard: BlobData,
+    content: Vec<BlobData>,
+}
+
+/// An incomplete bao file consists of outboard and content.
+///
+/// They are related to each other.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct IncompleteBaoFile {
+    outboard: MutableBlobData,
+    content: MutableBlobData,
+}
+
+/// The data we retain for each hash
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+struct HashData {
```

Comment:
Is this not better as an enum?  Or am I misunderstanding this type?

Reply:
I wanted to leave the option to have both. But maybe that is not necessary. Why would you hold on to incomplete data if you have complete data.

Reply:
Hum, I don't think keeping incomplete data on disk after it's been completed would match user expectations.  You'd wonder where all your disk space is gone.  Bittorrent client also don't do this.  You know that as long as your client has this location and it's unmodified it can serve it, so it's up to you to manage it.  I think this works fine for iroh as well.